File: filter.py
EXCLUSION_CHAR = "^"
UNKNOWN_CHAR = "_"

FILTER_LENGTH = 5
EMPTY_FILTER = UNKNOWN_CHAR * FILTER_LENGTH

class Filter:
   def __init__(self, include_str: str, exclude_str: str):
      
      # Sanity check include string
      stripped_include = include_str.replace(UNKNOWN_CHAR, "")
      include_len = len(stripped_include)
      if include_len > 0:
         assert stripped_include.isalpha(), f"Filter include string '{include_str}' is not alphabetic"
      
      include_list = list(include_str)
      while len(include_list) < FILTER_LENGTH:
         include_list.append(UNKNOWN_CHAR)
      assert len(include_list) == FILTER_LENGTH, f"Filter include string '{include_str}' is not 5 characters long"

      # Sanity check exclude string
      stripped_exclude = exclude_str.replace(UNKNOWN_CHAR, "")
      exclude_len = len(stripped_exclude)
      if exclude_len > 0:
         assert stripped_exclude.isalpha(), f"Filter exclude string '{exclude_str}' is not alphabetic"
      
      exclude_list = list(exclude_str)
      while len(exclude_list) < FILTER_LENGTH:
         exclude_list.append(UNKNOWN_CHAR)
      assert len(exclude_list) == FILTER_LENGTH, f"Filter exclude string '{exclude_str}' is not 5 characters long"

      if include_len > 0 and exclude_len > 0:
         for i, c in enumerate(include_list):
            if c != UNKNOWN_CHAR:
               assert exclude_list[i] == UNKNOWN_CHAR, f"Filter exclude string '{exclude_str}' collides with include '{include_str}'"
            else:
               assert exclude_list[i] != UNKNOWN_CHAR, f"Filter exclude string '{exclude_str}' collides with include '{include_str}'"

      self.include = include_list
      self.exclude = exclude_list
      self.is_empty = include_len==0 and exclude_len==0

   @classmethod
   def make_filter(cls, filter_str: str):
      filter_split_list = filter_str.split(EXCLUSION_CHAR)

      if len(filter_split_list) == 2:
         return Filter(filter_split_list[0], filter_split_list[1])
      elif len(filter_split_list) == 1:
         return Filter(filter_split_list[0], EMPTY_FILTER)
      else:
         raise ValueError(f"Invalid filter string {filter_str}")

   @classmethod
   def make_filter_func(cls, filter_str: str):
      filter = Filter.make_filter(filter_str)

      def filter_func(maybe_sol: str) -> bool:
         assert maybe_sol.isalpha(), f"Possible solution {maybe_sol} is not alphabetic"
         assert maybe_sol.upper() == maybe_sol, f"Possible solution {maybe_sol} is not uppercase"
         assert len(maybe_sol) == 5, f"Possible solution {maybe_sol} is not 5 characters long"

         sol_list = list(maybe_sol)

         # Loop through the filter known positions
         for i, include_char in enumerate(filter.include):
            if include_char != UNKNOWN_CHAR and include_char.isupper():
               if sol_list[i] != include_char:
                  return False
               sol_list[i] = UNKNOWN_CHAR

         # Loop through the filter unknown positions
         for i, include_char in enumerate(filter.include):
            if include_char != UNKNOWN_CHAR and include_char.islower():
               up_ch = include_char.upper()
               if sol_list[i] == up_ch:
                  return False
             
               try:
                  pos = sol_list.index(up_ch)
               except:
                  return False
               sol_list[pos] = UNKNOWN_CHAR

         # Loop through the filter exclusions
         for i, exclude_char in enumerate(filter.exclude):
            if exclude_char != UNKNOWN_CHAR:
               up_ch = exclude_char.upper()
               if up_ch in sol_list:
                  return False

         return True
      return filter_func
   
   @classmethod
   def make_hint(cls, guess: str, solution: str) -> "Filter":
      sol = list(solution)
      # Make a hint based on the guess and the solution
      include = list(EMPTY_FILTER)
      exclude = list(EMPTY_FILTER)
      for i, c in enumerate(guess):
         if c == sol[i]:
            include[i] = c.upper()
            sol[i] = UNKNOWN_CHAR

      for i, c in enumerate(guess):
         try:
            pos = sol.index(c)
         except ValueError:
            if include[i] != c:
               exclude[i] = c.lower()
            continue
         include[i] = c.lower()
         sol[pos] = UNKNOWN_CHAR
 
      return Filter("".join(include), "".join(exclude))

   def __str__(self) -> str:
      filter_str = "".join(self.include)
      exclude_str = "".join(self.exclude)
      if exclude_str != EMPTY_FILTER:
         filter_str += EXCLUSION_CHAR + exclude_str
      return filter_str
   
   def is_match(self, guess_str: str) -> bool:
      return ("".join(self.include) == guess_str) and ("".join(self.exclude) == EMPTY_FILTER)
      
   @classmethod
   def unit_test_hint(cls):
      # Test the make_hint function
      test_cases = [
         ("ABACK", "BREAD", "ab___^__ack"),
         ("ABACK", "BREAK", "ab__K^__ac_"),
         ("ABACK", "BREAM", "ab___^__ack"),
         
         ("ABAMK", "BREAD", "ab___^__amk"),
         ("ABAMK", "BREAK", "ab__K^__am_"),
         ("ABAMK", "BREAM", "ab_m_^__a_k"),
      ]
      for guess, solution, expected in test_cases:
         hint = cls.make_hint(guess, solution)
         
         assert str(hint) == expected, f"Make hint {solution} failed for guess {guess}. Expected {expected}, got {hint}"

   @classmethod
   def unit_test_filter(cls):
      # Test the filter function
      test_cases = [
         ("BAC__", "BACON", True),
         
         ("_ALE", "VALET", True),
         ("_ALE", "PALER", True),
         ("_ALE", "BALER", True),
         ("_ALE", "PALED", True),
         
         ("_ALEb", "BALER", True),
         ("_ALEb", "BALED", True),
         
         ("_alek", "SALEK", False),
         ("_alek", "ALIKE", True),
         ("_alek", "ANKLE", True),
         ("_alek", "FLAKE", True),
         ("_alek", "LEAKY", True),
         ("_alek", "SLAKE", True),
         
         ("_kela", "SALEK", True),
         ("_kela", "ALIKE", True),
         ("_kela", "FLAKE", True),
         ("_kela", "LEAKY", True),
         ("_kela", "SLAKE", True),
         ("_kela", "LATKE", True),
         
         ("Alek", 'ANKLE', True),
         
         ("ab___^__ack", "BREAD", True),
         ("ab__K^__ac_", "BREAK", True),
         ("ab___^__ack", "BREAM", True),
         ("ab___^__ack", "ABACK", False),
         ("ab__K^__ac_", "ABACK", False),
         ("ab___^__ack", "ABACK", False),
                   
         ("ab___^__amk", "BREAD", True),
         ("ab__K^__am_", "BREAK", True),
         ("ab_m_^__a_k", "BREAM", True),
         
         ("BREA_^____d", "BREAD", False),
         ("BREA_^____d", "BREAK", True),
         ("BREA_^____d", "BREAM", True),
      ]
      for filter, candidate, expected in test_cases:
         filter_func = cls.make_filter_func(filter)
         result = filter_func(candidate)
         
         assert result == expected, f"Filter {filter} failed for candidate {candidate}. Expected {expected}, got {result}"

# Run unit tests
Filter.unit_test_filter()
Filter.unit_test_hint()

File: rust_solver/src/filter.rs
pub const UNKNOWN_CHAR: char = '_';
pub const FILTER_LENGTH: usize = 5;
pub const EMPTY_FILTER: &str = "_____";

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Filter {
    pub guess_str: String,
    pub feedback_str: String,

    pub include: Vec<char>,
    pub exclude: Vec<char>,
    pub is_empty: bool,
}

impl Filter {
    pub fn new(guess_str: &str, feedback_str: &str) -> Result<Self, String> {
        assert!(
            guess_str.len() == FILTER_LENGTH,
            "Filter guess string '{guess_str}' is not 5 characters!"
        );

        assert!(
            guess_str.chars().all(char::is_alphabetic),
            "Filter guess string '{guess_str}' is not alphabetic"
        );
        let guess = guess_str.as_bytes();

        let mut feedback = feedback_str.chars().collect::<Vec<_>>();
        feedback.resize(FILTER_LENGTH, UNKNOWN_CHAR);
        assert!(
            feedback.len() == FILTER_LENGTH,
            "Filter string '{feedback_str}' is not 5 characters long"
        );

        let mut include = vec![];
        let mut exclude: Vec<char> = vec![];
        for (i, feedback_char) in feedback.iter().enumerate() {
            let guess_char = guess[i] as char;
            if guess_char.is_ascii_uppercase() {
                include.push(guess_char);
                exclude.push(UNKNOWN_CHAR)
            } else if *feedback_char == UNKNOWN_CHAR {
                include.push(UNKNOWN_CHAR);
                exclude.push(guess_char.to_ascii_lowercase())
            } else {
                include.push(guess_char);
                exclude.push(UNKNOWN_CHAR)
            }
        }

        let include_str: String = include.iter().collect();

        Ok(Self {
            guess_str: guess_str.to_string(),
            feedback_str: feedback_str.to_string(),
            include,
            exclude,
            is_empty: include_str == EMPTY_FILTER,
        })
    }

    pub fn make_filter(filter_str: &str) -> Result<Self, String> {
        assert!(
            filter_str.len() >= FILTER_LENGTH,
            "Filter string '{filter_str}' is less than 5 characters!"
        );

        let guess_str = &filter_str[..FILTER_LENGTH];
        let feedback = &filter_str[FILTER_LENGTH..];

        Self::new(guess_str, feedback)
    }

    pub fn make_filter_func(&self) -> impl Fn(&str) -> bool + '_ {
        move |maybe_sol| {
            assert!(
                maybe_sol.chars().all(char::is_alphabetic),
                "Possible solution {maybe_sol} is not alphabetic"
            );
            assert!(
                maybe_sol.chars().all(char::is_uppercase),
                "Possible solution {maybe_sol} is not uppercase"
            );
            assert_eq!(
                maybe_sol.chars().count(),
                FILTER_LENGTH,
                "Possible solution {maybe_sol} is not 5 characters long"
            );

            let mut solution = maybe_sol.chars().collect::<Vec<_>>();

            for (index, &include_char) in self.include.iter().enumerate() {
                if include_char != UNKNOWN_CHAR && include_char.is_uppercase() {
                    if solution[index] != include_char {
                        return false;
                    }
                    solution[index] = UNKNOWN_CHAR;
                }
            }

            for (index, &include_char) in self.include.iter().enumerate() {
                if include_char != UNKNOWN_CHAR && include_char.is_lowercase() {
                    let upper = include_char.to_uppercase().next().unwrap();
                    if solution[index] == upper {
                        return false;
                    }

                    let Some(position) = solution.iter().position(|&character| character == upper)
                    else {
                        return false;
                    };
                    solution[position] = UNKNOWN_CHAR;
                }
            }

            for &exclude_char in &self.exclude {
                if exclude_char != UNKNOWN_CHAR {
                    let upper = exclude_char.to_uppercase().next().unwrap();
                    if solution.contains(&upper) {
                        return false;
                    }
                }
            }

            true
        }
    }

    pub fn make_hint(guess_str: &str, solution_str: &str) -> Result<Self, String> {
        let guess = guess_str.to_uppercase();
        assert!(
            guess.chars().all(char::is_alphabetic),
            "Guess string '{guess}' is not alphabetic!"
        );
        assert!(
            guess.chars().count() == FILTER_LENGTH,
            "Guess string '{guess}' is not 5 characters long!"
        );

        let solution_str = solution_str.to_uppercase();
        assert!(
            solution_str.chars().all(char::is_alphabetic),
            "Solution string '{solution_str}' is not alphabetic!"
        );
        assert!(
            solution_str.chars().count() == FILTER_LENGTH,
            "Solution string '{solution_str}' is not 5 characters long!"
        );

        let guess_chars: Vec<char> = guess.chars().collect();
        let solution: Vec<char> = solution_str.chars().collect();
        let mut remaining = solution.clone();
        let mut guess_out = vec![];
        let mut feedback = vec![];

        for (index, guess_char) in guess_chars.iter().enumerate() {
            if *guess_char == solution[index] {
                remaining[index] = UNKNOWN_CHAR;
                guess_out.push(*guess_char);
                feedback.push(*guess_char);
            } else {
                guess_out.push(UNKNOWN_CHAR);
                feedback.push(UNKNOWN_CHAR);
            }
        }

        for (index, guess_char) in guess_chars.iter().enumerate() {
            if guess_out[index] == UNKNOWN_CHAR {
                if let Some(pos) = remaining.iter().position(|&b| b == *guess_char) {
                    remaining[pos] = UNKNOWN_CHAR;
                    feedback[index] = guess_char.to_ascii_lowercase();
                }
                guess_out[index] = guess_char.to_ascii_lowercase();
            }
        }

        let guess_out: String = guess_out.into_iter().collect();
        let feedback: String = feedback.into_iter().collect();

        Self::new(&guess_out, &feedback)
    }
}

impl std::fmt::Display for Filter {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.feedback_str != EMPTY_FILTER {
            write!(formatter, "{}{}", self.guess_str, self.feedback_str)
        } else {
            write!(formatter, "{}", self.guess_str)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hint_matches_python_cases() {
        let test_cases = [
            ("ABACK", "BREAD", "abackab___"),
            ("ABACK", "BREAK", "abacKab__K"),
            ("ABACK", "BREAM", "abackab___"),
            ("ABAMK", "BREAD", "abamkab___"),
            ("ABAMK", "BREAK", "abamKab__K"),
            ("ABAMK", "BREAM", "abamkab_m_"),
            ("WHOSE", "WORSE", "WhoSEW_oSE"),
            ("WORSE", "WHOSE", "WorSEWo_SE"),
        ];

        for (guess, solution, expected) in test_cases {
            assert_eq!(
                Filter::make_hint(guess, solution).unwrap().to_string(),
                expected
            );
        }
    }

    #[test]
    fn filter_matches_python_cases() {
        let test_cases = [
            ("BACxx", "BACON", true),
            ("xALEx", "VALET", true),
            ("xALEx", "PALER", true),
            ("xALEx", "BALER", true),
            ("xALEx", "PALED", true),
            ("xALEb____b", "BALER", true),
            ("xALEb____b", "BALED", true),
            ("xalek_alek", "SALEK", false),
            ("xalek_alek", "ALIKE", true),
            ("xalek_alek", "ANKLE", true),
            ("xalek_alek", "FLAKE", true),
            ("xalek_alek", "LEAKY", true),
            ("xalek_alek", "SLAKE", true),
            ("xkela_kela", "SALEK", true),
            ("xkela_kela", "ALIKE", true),
            ("xkela_kela", "FLAKE", true),
            ("xkela_kela", "LEAKY", true),
            ("xkela_kela", "SLAKE", true),
            ("xkela_kela", "LATKE", true),
            ("AlekxAlek", "ANKLE", true),
            ("abackab", "BREAD", true),
            ("abacKab", "BREAK", true),
            ("abackab", "BREAM", true),
            ("abacKab", "ABACK", false),
            ("abacK__ac_", "ABACK", false),
            ("abamkab", "BREAD", true),
            ("abamKab", "BREAK", true),
            ("abamkab_m", "BREAM", true),
            ("BREAd____d", "BREAD", false),
            ("BREAd", "BREAK", true),
            ("BREAd", "BREAM", true),
            ("WhoSE", "WORSE", false),
            ("WhoSE__o", "WORSE", true),
            ("WorSE", "WHOSE", false),
            ("WorSE_o", "WHOSE", true),
        ];

        for (filter_str, candidate, expected) in test_cases {
            let filter = Filter::make_filter(filter_str).unwrap();
            eprintln!("Filter: {filter_str}, Candidate: {candidate}, Expected: {expected}");
            assert_eq!(
                filter.make_filter_func()(candidate),
                expected,
                "Filter {filter_str}"
            );
        }
    }
}
